gh-126615: Make `COMError` public and add to `ctypes` doc. (#126686)
python/cpython: Doc/library/ctypes.rst

   integer. *name* is name of the COM method. *iid* is an optional pointer to
   the interface identifier which is used in extended error reporting.

>    If *iid* is not specified, a :exc:`WindowsError` is raised if the COM method

Reviewer: [`WindowsError`](https://docs.python.org/3/library/exceptions.html#WindowsError) is now an alias of `OSError`.
```suggestion
   If *iid* is not specified, an :exc:`OSError` is raised if the COM method
```

Author: Good catch.

Thank you.

>    call fails. If *iid* is specified, a :exc:`.COMError` is raised instead.

Reviewer: I think (but I'm not 100% sure) that `.COMError` means Sphinx will search *all* namespaces, not just `ctypes`, for a `COMError`. So it would be better to use `ctypes.COMError` explicitly:
```suggestion
   call fails. If *iid* is specified, a :exc:`~ctypes.COMError` is raised instead.
```

   integer. *name* is name of the COM method. *iid* is an optional pointer to
   the interface identifier which is used in extended error reporting.

   If *iid* is not specified, a :exc:`WindowsError` is raised if the COM method

>    call fails. If *iid* is specified, a :exc:`.COMError` is raised instead.

Reviewer 2: ```suggestion
   call fails. If *iid* is specified, a :exc:`COMError` is raised instead.
```

   COM methods use a special calling convention: They require a pointer to
   the COM interface as first argument, in addition to those parameters that
   are specified in the :attr:`!argtypes` tuple.

> .. exception:: COMError(hresult, text, details)

Reviewer 2: Should it be placed here or somewhere else? because it cuts the flow of the reading where *paramflags* is documented afterwards. Maybe we can put it *before* the functions?

Author: I was also unsure whether this was the right place to put it.

On the other hand, placing it above `prototype` or above `FOOFUNCTYPE` might seem abrupt.

Might it be appropriate to create the `Exceptions` section, including `ArgumentError`?
However, in this scope, I think making such a change that involves the "Foreign functions" section is excessive.

Reviewer 2: Well.. honestly I think it makes sense to put it in an exception section. I think it could be fine for this small docs change (I mean, it's for our own convenience). We can ask @hugovk as a docs expert.

Reviewer 3: Yes, we often have an exceptions section further down, see for example:

https://docs.python.org/3/library/json.html#exceptions

Author: If the community agrees, creating an "Exceptions" section is no problem at all.

For now, I think I'll try placing it at the end of the document, after the "Arrays and pointers" section.

>    Windows only: This non-public exception is raised when a COM method call

Reviewer 2: You can use `.. availability:: Windows` (but put the directive after the class doc)

Author: I followed the pattern in `ctypes.rst` because it frequently used sentences like `Windows only: ...`.

Since this markup wasn't mentioned in [the documentation guide](https://devguide.python.org/documentation/markup/#), I overlooked it.
Is this a more modern approach?

Reviewer 2: It's widely used in the https://docs.python.org/3/library/socket.html module actually. But if it's not the pattern of `ctypes` maybe it's better to be consistent. We can make a follow-up PR to transform them into directives.

Author: I removed the `Availability` directive and reverted to the original "Windows only: ...".

Once this PR is merged, I would like to work for replacing "Foo only: ..." with the `.. availability:: Foo` directive.

Author: Referring to the [`socket.ioctl`](https://docs.python.org/3/library/socket.html#socket.socket.ioctl) documentation, it seems that using `:platform: Windows` might be more appropriate than `.. availability:: Windows`. What do you think?

Alternatively, could it be that `:platform:` is intended to be used exclusively in module sections, as described in the [documentation guide's "Module-specific markup"](https://devguide.python.org/documentation/markup/#module-specific-markup)?

Reviewer: `.. availability::` is generally the correct one to use. FWIW, we have some [tooling](https://github.com/python/cpython/blob/main/Doc/tools/extensions/availability.py) to ensure it's in a correct format.
It's OK to not use it in this PR, to be consistent with the rest of the documentation. But, if you're looking for more to do in the ctypes docs, adding `availability` would make a great follow-up PR :)

   failed.

   .. attribute:: hresult

      The integer value representing the error code.

   .. attribute:: text

      The error message.

   .. attribute:: details

      The 5-tuple representing additional details about the error.

> 
> The optional *paramflags* parameter creates foreign function wrappers with much

Reviewer 2: ```suggestion

The optional *paramflags* parameter creates foreign function wrappers with much
```

   integer. *name* is name of the COM method. *iid* is an optional pointer to
   the interface identifier which is used in extended error reporting.

   If *iid* is not specified, a :exc:`WindowsError` is raised if the COM method

>    call fails. If *iid* is specified, a :exc:`COMError` is raised instead.

Reviewer 2: Err... I wonder why the reference is not picked up here, because the class *is* documented. (May be a Sphinx issue?) Sorry, but you'll need the `.COMError` as before (my bad...)

Author: It has become harder to see after the rebase, but in the earlier [commit without the dot](https://github.com/python/cpython/commit/371110d4959b399ec87b859365778b131999b51b), "Tests / Docs / Docs (pull_request)" failed in CI.
In contrast, the [commit with the dot](https://github.com/python/cpython/commit/58cc032eb75a08ad64ea58317ce2ef93bc5fc34c) succeeded. 

Additionally, without the dot, no link was generated.

I think also it's probably an issue with Sphinx.

Reviewer 2: > Additionally, without the dot, no link was generated.

That's because it couldn't find the reference. I'm wondering whether it couldn't find the reference because the exception is documented after or because of how it's declared (it might also happen that the module's scope was changed somewhere)

        Returns the object to which to pointer points.  Assigning to this
        attribute changes the pointer to point to the assigned object.

.. _ctypes-exceptions:

Exceptions
^^^^^^^^^^

.. exception:: ArgumentError

   This exception is raised when a foreign function call cannot convert one of the
   passed arguments.

.. exception:: COMError(hresult, text, details)

   Windows only: This non-public exception is raised when a COM method call
   failed.

   .. attribute:: hresult

      The integer value representing the error code.

   .. attribute:: text

      The error message.

   .. attribute:: details

>       The 5-tuple representing additional details about the error.

Reviewer: That's not very helpful. Could you document what the items are?

Author: Based on [the C implementation source code](https://github.com/python/cpython/blob/f9c5573dedcb2f2e9ae152672ce157987cdea612/Modules/_ctypes/callproc.c#L1098-L1164), these items are retrieved via methods of [`IErrorInfo`](https://learn.microsoft.com/en-us/windows/win32/api/oaidl/nn-oaidl-ierrorinfo)[^1]: [`GetDescription`](https://learn.microsoft.com/en-us/windows/win32/api/oaidl/nf-oaidl-ierrorinfo-getdescription), [`GetSource`](https://learn.microsoft.com/en-us/windows/win32/api/oaidl/nf-oaidl-ierrorinfo-getsource), [`GetHelpFile`](https://learn.microsoft.com/en-us/windows/win32/api/oaidl/nf-oaidl-ierrorinfo-gethelpfile), [`GetHelpContext`](https://learn.microsoft.com/en-us/windows/win32/api/oaidl/nf-oaidl-ierrorinfo-gethelpcontext), and the ProgID obtained via using [`ProgIDFromCLSID`](https://learn.microsoft.com/en-us/windows/win32/api/combaseapi/nf-combaseapi-progidfromclsid) and [`GetGUID`](https://learn.microsoft.com/en-us/windows/win32/api/oaidl/nf-oaidl-ierrorinfo-getguid).

However, this exception can be raised not only from foreign functions but also intentionally by passing variable-length tuples, as seen in `test_win32.py` and the `comtypes` package.
https://github.com/python/cpython/blob/f9c5573dedcb2f2e9ae152672ce157987cdea612/Lib/test/test_ctypes/test_win32.py#L67-L76

I will revise this part, taking these facts into account.

[^1]: Reference: [`IErrorInfo` implementation in `comtypes`](https://github.com/enthought/comtypes/blob/1ea74f5a68c4f1e0ee511ba3591e2204688ec071/comtypes/errorinfo.py#L21-L35).

Reviewer: Since `_ctypes` is internal API that is subject to change, I think it would be best to change `test_win32` and `comtypes` so that they raise the same 5-tuple as `ctypes` itself, and document that 5-tuple.

Author: I wrote details of `details`.

I plan to improve `test_win32` after improving this documentation.

For `comtypes`, I will investigate how much `COMError.details` is being used and then discuss the matter with the project's community to decide how to proceed.

As mentioned earlier, I would like to focus on improving the documentation in this PR.

        Returns the object to which to pointer points.  Assigning to this
        attribute changes the pointer to point to the assigned object.

.. _ctypes-exceptions:

Exceptions
^^^^^^^^^^

.. exception:: ArgumentError

   This exception is raised when a foreign function call cannot convert one of the
   passed arguments.

.. exception:: COMError(hresult, text, details)

   Windows only: This exception is raised when a COM method call failed.

   .. attribute:: hresult

      The integer value representing the error code.

   .. attribute:: text

      The error message.

   .. attribute:: details

      The 5-tuple ``(descr, source, helpfile, helpcontext, progid)``.

      *descr* is the textual description.  *source* is the language-dependent
      ``ProgID`` for the class or application that raised the error.  *helpfile*
      is the path of the help file.  *helpcontext* is the help context
      identifier.  *progid* is the ``ProgID`` of the interface that defined the
      error.

>    .. versionchanged:: 3.14
> 
>       This exception is now public.
>       Previously, this was private and only available in ``_ctypes``.

Reviewer: Let's not expose private API in the docs.

```suggestion
   .. versionadded:: next
```

Author: You’re right, this information isn’t particularly useful in the documentation.

I’ll move it to the NEWS file as a supplement.
